src/handshake.rs

```diff
@@ -0,0 +1,171 @@
+            AllowStd<S>,
+        ) -> Result<Role::FinalResult, Error<Role>>
+        + Unpin,
+        S: AsyncRead + AsyncWrite + Unpin,
+{
+    let start = StartedHandshakeFuture(Some(StartedHandshakeFutureInner { f, stream }));
+
+    match start.await? {
+        StartedHandshake::Done(r) => Ok(r),
+        StartedHandshake::Mid(s) => {
+            let res: Result<Role::FinalResult, Error<Role>> = MidHandshake::<Role>(Some(s)).await;
+            res
+        },
+    }
+}
+
+pub(crate) async fn client_handshake<F, S>(stream: S, f: F) -> Result<(WebSocketStream<S>, Response), Error<ClientHandshake<AllowStd<S>>>>
+    where
+        F: FnOnce(
+            AllowStd<S>,
+        ) -> Result< <ClientHandshake<AllowStd<S>> as HandshakeRole>::FinalResult, Error<ClientHandshake<AllowStd<S>>>>
+        + Unpin,
+        S: AsyncRead + AsyncWrite + Unpin,
+{
+    let result = handshake(stream, f).await?;
+    let (s, r) = result;
+    Ok( (WebSocketStream::new(s), r) )
+}
+
+pub(crate) async fn server_handshake<C, F, S>(stream: S, f: F) -> Result<WebSocketStream<S>, Error<ServerHandshake<AllowStd<S>, C>>>
+    where
+        C: Callback + Unpin,
+        F: FnOnce(
+            AllowStd<S>,
+        ) -> Result< <ServerHandshake<AllowStd<S>, C> as HandshakeRole>::FinalResult, Error<ServerHandshake<AllowStd<S>, C>>>
+        + Unpin,
+        S: AsyncRead + AsyncWrite + Unpin,
+{
+    let s: WebSocket<AllowStd<S>> = handshake(stream, f).await?;
+    Ok(WebSocketStream::new(s))
+}
+
+impl<Role, F, S> Future for StartedHandshakeFuture<F, S>
+    where
+        Role: HandshakeRole,
+        F: FnOnce(
+            AllowStd<S>,
+        ) -> Result<Role::FinalResult, Error<Role>> + Unpin,
+        S: Unpin,
+        AllowStd<S>: Read + Write,
+{
+    type Output = Result<StartedHandshake<Role>, Error<Role>>;
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        ctx: &mut Context<'_>,
+    ) -> Poll<Self::Output> {
+        let inner = self.0.take().expect("future polled after completion");
+        let stream = AllowStd {
+            inner: inner.stream,
+            context: ctx as *mut _ as *mut (),
+        };
+
+        match (inner.f)(stream) {
+            Ok(r) => {
+                Poll::Ready(Ok(StartedHandshake::Done(r)))
+            }
+            Err(Error::Interrupted(mid)) => {
+                Poll::Ready(Ok(StartedHandshake::Mid(mid)))
+            }
+            Err(Error::Failure(e)) => Poll::Ready(Err(Error::Failure(e))),
+        }
+    }
+}
+
+impl<Role> Future for MidHandshake<Role>
+    where
+        Role: HandshakeRole + Unpin,
+{
+    type Output = Result<Role::FinalResult, Error<Role>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut_self = self.get_mut();
+        let mut s = mut_self.0.take().expect("future polled after completion");
+
+        let machine = s.get_mut();
+
+        (machine.get_mut() as AllowStd<_>).context = cx as *mut _ as *mut ();
```

Comment:
Currently blocked here. Need to figure out how to get the context back out of the machine to set it appropriately.